File: Part 2 demos/matlab/ID_test.m
function ID_test

Npan      = 200;  % The number of Gaussian panels on each smooth component.
Ngau      = 10;    % The number of Gaussian nodes on each panel.

acc       = 1e-8; % The tolerance in the HSS compression.


[ww,C,tt,t_sta,t_end] = construct_geom(Npan,Ngau);
Ntot = size(C,2);

ind = 1751:Ntot;

indoffd = 1:1750;

Aall = LOCAL_construct_A_offd(C,ww,ind,indoffd);


[U,J,k] = constructID(Aall,acc);

err = norm(Aall-U*Aall(J(1:k),:));

fprintf(1,'err = %12.5e\n', err)

plot(C(1,ind),C(4,ind),'r-','LineWidth',4)
hold on
plot(C(1,indoffd),C(4,indoffd),'b-','LineWidth',4)
plot(C(1,ind(J(1:k))),C(4,ind(J(1:k))),'cx','LineWidth',2)

return
end

function [U,J,k] = constructID(Amat,acc)

f = 1.2;
[U,J] = ID(Amat,'tol', acc, f);
k = size(U,2);

return
end

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
function [ww,C,tt,t_sta,t_end] = construct_geom(Npan,Ngau)

t_sta         = 0;    % Start point of the discretization.
t_end         = 1; % End   point of the discretization.
tt_panels     = linspace(t_sta,t_end,Npan+1);
[tt,ww_gauss] = LOCAL_get_gauss_nodes(tt_panels,Ngau);

C             = zeros(6,length(tt));
C(1,:)        = (1+0.3*cos(5*2*pi*tt)).*cos(2*pi*tt);
C(2,:)        = 2*pi*(-sin(2*pi*tt).*(0.3*cos(5*2*pi*tt)+1)-1.5*sin(5*2*pi*tt).*cos(2*pi*tt));

C(3,:)        =  (2*pi)^2*(3*sin(2*pi*tt).*sin(5*2*pi*tt)+cos(2*pi*tt).*(-7.8*cos(5*2*pi*tt)-1));
C(4,:)        =  sin(2*pi*tt).*(1+0.3*cos(5*2*pi*tt));
C(5,:)        =  2*pi*(-1.5*sin(2*pi*tt).*sin(5*2*pi*tt)+0.3*cos(5*2*pi*tt).*cos(2*pi*tt)+cos(2*pi*tt));
C(6,:)        = (2*pi)^2*(sin(2*pi*tt).*(-7.8*cos(5*2*pi*tt)-1)-3*sin(5*2*pi*tt).*cos(2*pi*tt));

ww            = ww_gauss.*sqrt(C(2,:).*C(2,:) + C(5,:).*C(5,:));
return
end

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

function B = LOCAL_construct_A_offd(C,ww,ind1,ind2)

n1   = length(ind1);
n2   = length(ind2);
dd1  = C(1,ind1)' * ones(1,n2) - ones(n1,1) * C(1,ind2);
dd2  = C(4,ind1)' * ones(1,n2) - ones(n1,1) * C(4,ind2);
ddsq = dd1.*dd1 + dd2.*dd2;
nn1  = ones(n1,1)*( C(5,ind2)./sqrt(C(2,ind2).*C(2,ind2) + C(5,ind2).*C(5,ind2)));
nn2  = ones(n1,1)*(-C(2,ind2)./sqrt(C(2,ind2).*C(2,ind2) + C(5,ind2).*C(5,ind2)));
B    = -(1/(2*pi))*(nn1.*dd1 + nn2.*dd2)./ddsq;
B    = B.*(ones(n1,1)*ww(ind2));

return
end
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

function [tt,ww] = LOCAL_get_gauss_nodes(tt_panels,Ngau)

npanels = length(tt_panels)-1;
tt   = zeros(1,npanels*Ngau);
ww   = zeros(1,npanels*Ngau);
[t_ref,w_ref] = LOCAL_lgwt(Ngau,0,1);
t_ref         = t_ref(end:(-1):1);
w_ref         = w_ref(end:(-1):1);
for i = 1:npanels
    h       = tt_panels(i+1) - tt_panels(i);
    ind     = (i-1)*Ngau + (1:Ngau);
    tt(ind) = tt_panels(i) + h*t_ref;
    ww(ind) = h*w_ref;
end

return
end


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

function [x,w] = LOCAL_lgwt(N,a,b)

% lgwt.m
%
% This script is for computiNgau definite integrals usiNgau Legendre-Gauss
% Quadrature. Computes the Legendre-Gauss nodes and weights  on an interval
% [a,b] with truncation order N
%
% Suppose you have a continuous function f(x) which is defined on [a,b]
% which you can evaluate at any x in [a,b]. Simply evaluate it at all of
% the values contained in the x vector to obtain a vector f. Then compute
% the definite integral usiNgau sum(f.*w);
%
% Written by Greg von Winckel - 02/25/2004
N=N-1;
N1=N+1; N2=N+2;

xu=linspace(-1,1,N1)';

% Initial guess
y=cos((2*(0:N)'+1)*pi/(2*N+2))+(0.27/N1)*sin(pi*xu*N/N2);

% Legendre-Gauss Vandermonde Matrix
L=zeros(N1,N2);

% Derivative of LGVM
Lp=zeros(N1,N2);

% Compute the zeros of the N+1 Legendre Polynomial
% usiNgau the recursion relation and the Newton-Raphson method

y0=2;

% Iterate until new points are uniformly within epsilon of old points
while max(abs(y-y0))>eps


    L(:,1)=1;
    Lp(:,1)=0;

    L(:,2)=y;
    Lp(:,2)=1;

    for k=2:N1
        L(:,k+1)=( (2*k-1)*y.*L(:,k)-(k-1)*L(:,k-1) )/k;
    end

    Lp=(N2)*( L(:,N1)-y.*L(:,N2) )./(1-y.^2);

    y0=y;
    y=y0-L(:,N2)./Lp;

end

% Linear map from[-1,1] to [a,b]
x=(a*(1-y)+b*(1+y))/2;

% Compute the weights
w=(b-a)./((1-y.^2).*Lp.^2)*(N2/N1)^2;

return
end

File: Part 2 demos/python/ID_test.py
import numpy as np
import matplotlib.pyplot as plt
import scipy.linalg.interpolative as sli
from numpy import linalg as LA

def ID_test():

#  The number of Gaussian panels on each smooth component.
     Npan      = 200
# The number of Gaussian nodes on each panel.     
     Ngau      = 10    
     # The tolerance in the HSS compression.
     acc       = 1e-8

# create geometry
     [C,ww,tt] = construct_geom(Npan,Ngau)
      
# create indices of interest
     ind = np.arange(1750,1999,1)
     indoffd = np.arange(0,1749,1)

# create matrix we are trying to approximate         
     Aall = construct_A_offd(C,ww,ind,indoffd)
     
# create the ID factorization
     [k, J, proj] = sli.interp_decomp(np.transpose(Aall), acc)
     
# validate the result
     B = sli.reconstruct_skel_matrix(np.transpose(Aall), k, J)
     B = np.transpose(B)
     P = sli.reconstruct_interp_matrix(J, proj)
     P = np.transpose(P)
     
     err = LA.norm(Aall-np.dot(P,B))
     print('err is:', err)
          
# plot the geometry          
     plt.figure()    
     plt.plot(C[0,ind],C[3,ind],'r-')
     plt.plot(C[0,indoffd],C[3,indoffd],'b-')
     plt.plot(C[0,ind[J[1:k]]],C[3,ind[J[1:k]]],'cx')
     plt.show()
        
     return 

def construct_A_offd(C,ww,ind1,ind2):

     n1   = int(len(ind1))
     n2   = int(len(ind2))
     dd1  = np.outer(C[0,ind1] ,np.ones((1,n2))) - \
            np.outer(np.ones((1,n1)), C[0,ind2])
     dd2  = np.outer(C[3,ind1],np.ones((1,n2))) - \
            np.outer(np.ones((1,n1)), C[3,ind2])
     ddsq = dd1*dd1 + dd2*dd2
     tmp1 = C[4,ind2]/np.sqrt(C[1,ind2]*C[1,ind2] + C[4,ind2]*C[4,ind2])
     nn1  = np.outer(np.ones((1,n1)),tmp1)
     tmp2 = C[1,ind2]/np.sqrt(C[1,ind2]*C[1,ind2] + C[4,ind2]*C[4,ind2])
     nn2  = np.outer(np.ones((1,n1)),-tmp2)
     B    = -(1/(2*np.pi))*(nn1*dd1 + nn2*dd2)/ddsq
     B    = B*np.outer(np.ones((1,n1)),ww[ind2])

     return B



def  construct_geom(Npan,Ngau):

     [tt,ww] = get_gauss_nodes(Ngau,Npan)
     ntot = int(Npan*Ngau)
     C = np.zeros((6,ntot))
     C[0,:] = (1+0.3*np.cos(5*2*np.pi*tt))*np.cos(2*np.pi*tt)
     C[1,:] = 2*np.pi*(-np.sin(2*np.pi*tt)*(0.3*np.cos(5*2*np.pi*tt)+1)- \
               1.5*np.sin(5*2*np.pi*tt)*np.cos(2*np.pi*tt))
     C[2,:] =  (2*np.pi)**2*(3*np.sin(2*np.pi*tt)*np.sin(5*2*np.pi*tt)+ \
               np.cos(2*np.pi*tt)*(-7.8*np.cos(5*2*np.pi*tt)-1))
     C[3,:] =  np.sin(2*np.pi*tt)*(1+0.3*np.cos(5*2*np.pi*tt))
     C[4,:] =  2*np.pi*(-1.5*np.sin(2*np.pi*tt)*np.sin(5*2*np.pi*tt) \
               +0.3*np.cos(5*2*np.pi*tt)*np.cos(2*np.pi*tt)+np.cos(2*np.pi*tt))
     C[5,:]  = (2*np.pi)**2*(np.sin(2*np.pi*tt)*(-7.8*np.cos(5*2*np.pi*tt)-1) \
               -3*np.sin(5*2*np.pi*tt)*np.cos(2*np.pi*tt))

     ww = ww*np.sqrt(C[1,:]*C[1,:] + C[4,:]*C[4,:])

     return [C,ww,tt]


def get_gauss_nodes(Ngau,Npan):

     tsta         = 0 
     tend         = 1 
     h = (tend-tsta)/Npan
     ttpanels = np.linspace(tsta,tend,Npan+1)
     tt = np.zeros(Npan*Ngau)
     ww = np.zeros(Npan*Ngau)
# reference quad     
     [tref,wref] = lgwt(Ngau,0,1)
     
     nvec = np.arange(0,Ngau)

     for j in range(Npan):
       h       = ttpanels[j+1] - ttpanels[j]
       ind     = j*Ngau + nvec
       tt[ind] = ttpanels[j] + h*np.flip(tref)
       ww[ind] = h*np.flip(wref)

     return[tt,ww]

# gauss legendre quadrature nodes and weights via
# Newton iteration


def lgwt(N,a,b):
  """ 
   This script is for computing definite integrals using Legendre-Gauss 
   Quadrature. Computes the Legendre-Gauss nodes and weights  on an interval
   [a,b] with truncation order N
  
   Suppose you have a continuous function f(x) which is defined on [a,b]
   which you can evaluate at any x in [a,b]. Simply evaluate it at all of
   the values contained in the x vector to obtain a vector f. Then compute
   the definite integral using np.sum(f*w)
  
   Written by Greg von Winckel - 02/25/2004
   translated to Python - 10/30/2022
  """
  N = N-1
  N1 = N+1
  N2 = N+2
  eps = np.finfo(float).eps  
  xu = np.linspace(-1,1,N1)
  
  # Initial guess
  y = np.cos((2*np.arange(0,N1)+1)*np.pi/(2*N+2))+(0.27/N1)*np.sin(np.pi*xu*N/N2)

  # Legendre-Gauss Vandermonde Matrix
  L = np.zeros((N1,N2))
  
  # Compute the zeros of the N+1 Legendre Polynomial
  # using the recursion relation and the Newton-Raphson method
  
  y0 = 2.
  one = np.ones((N1,))
  zero = np.zeros((N1,))

  # Iterate until new points are uniformly within epsilon of old points
  while np.max(np.abs(y-y0)) > eps:
      
    L[:,0] = one
    
    L[:,1] = y
    for k in range(2,N1+1): 
      L[:,k] = ((2*k-1)*y*L[:,k-1]-(k-1)*L[:,k-2])/k
    
    lp = N2*(L[:,N1-1]-y*L[:,N2-1])/(1-y**2)   
    
    y0 = y
    y = y0-L[:,N2-1]/lp
    
    
  
  # Linear map from[-1,1] to [a,b]
  x=(a*(1-y)+b*(1+y))/2
  
  # Compute the weights
  w=(b-a)/((1-y**2)*lp**2)*(N2/N1)**2
  return [x,w]
  
  
  


ID_test()
